voltixApp/VoltixApp/Services/Ethereum/EtherscanService.swift

```diff
@@ -1,132 +1,256 @@
+			async let gasPrice = fetchGasPrice()
+			async let nonce = fetchNonce(address: tx.fromAddress)
+			async let cryptoPrice = CryptoPriceService.shared.cryptoPrices?.prices[tx.coin.priceProviderId]?["usd"]
+			
+			if let priceRateUsd = await cryptoPrice {
+				tx.coin.priceRate = priceRateUsd
+			}
+			
+			if tx.coin.isToken {
+				tx.coin.rawBalance = try await fetchTokenRawBalance(contractAddress: tx.coin.contractAddress, address: tx.fromAddress)
+			} else {
+				tx.coin.rawBalance = try await fetchEthRawBalance(address: tx.fromAddress)
+			}
+			
+			tx.gas = String(try await gasPrice)
+			tx.nonce = try await nonce
+			
+			print("\(tx.coin.ticker) raw balance: \(tx.coin.rawBalance)")
+			print("\(tx.coin.ticker) GAS: \(tx.gas)")
+			print("\(tx.coin.ticker) Nonce: \(tx.nonce)")
+			
+		} catch let error as EtherScanError {
+			handleEtherScanError(error)
+		} catch {
+			print(error.localizedDescription)
+		}
+	}
+	
+	func broadcastTransaction(hex: String) async throws -> String {
+		let data = try await Utils.asyncPostRequest(urlString: Endpoint.broadcastEtherscanTransaction(hex: hex), headers: [:], body: Data())
+		guard let result = extractResult(fromData: data) else {
+			throw EtherScanError.resultExtractionFailed
+		}
+		return result
+	}
+	
+	func fetchTransactions(forAddress address: String) async throws -> ([EtherscanAPITransactionDetail], String) {
+		let decodedResponse: EtherscanAPIResponse = try await Utils.fetchObject(from: Endpoint.fetchEtherscanTransactions(address: address))
+		if let transactions = decodedResponse.result {
+			return (transactions, address)
+		} else {
+			throw EtherScanError.decodingError("Error to decode the transaction")
+		}
+	}
+	
+	func fetchERC20Transactions(forAddress address: String, contractAddress: String) async throws -> ([EtherscanAPITransactionDetail], String) {
+		let decodedResponse:EtherscanAPIResponse = try await Utils.fetchObject(from: Endpoint.fetchERC20Transactions(address: address, contractAddress: contractAddress))
+		if let transactions = decodedResponse.result {
+			return (transactions, address)
+		} else {
+			throw EtherScanError.decodingError("Error to decode the transaction")
+		}
+	}
+	
+	func estimateGasForERC20Transfer(tx: SendTransaction) async throws -> BigInt {
+		let data = constructERC20TransferData(recipientAddress: tx.toAddress, value: tx.amountInTokenWei)
+		let urlString = Endpoint.fetchEtherscanEstimateGasForERC20Transaction(data: data, contractAddress: tx.coin.contractAddress)
+		let resultData = try await Utils.asyncGetRequest(urlString: urlString, headers: [:])
+		
+		guard let resultString = extractResult(fromData: resultData) else {
+			throw EtherScanError.jsonDecodingError
+		}
+		
+		let trimmedResultString = resultString.trimmingCharacters(in: CharacterSet(charactersIn: "0x"))
+		guard let intResult = BigInt(trimmedResultString, radix: 16) else {
+			throw EtherScanError.conversionError
+		}
+		
+		return intResult
```

Comment:
The estimate gas is just an estimation , it doesn't guarantee the transaction will go through , in a lot cases , it might not ,  so it is better to inflate the result a little bit , for example , intResult * 1.5 etc